server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationConstraints.java

```diff
@@ -0,0 +1,124 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Predicate;
+
+/**
+ * Allocation constraints specify conditions which, if breached, reduce the priority of a node
+ * for receiving shard allocations.
+ */
+public class AllocationConstraints {
+    public final long CONSTRAINT_WEIGHT = 1000000L;
+    public static final Setting<ConstraintMode> INDEX_SHARDS_PER_NODE_CONSTRAINT_SETTING =
+            new Setting<>("aes.cluster.routing.allocation.constraint.index_shards_per_node", ConstraintMode.NONE.toString(),
+                    ConstraintMode::parse, Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    private final ConstraintMode indexShardsPerNodeConstraint;
+    private List<Predicate<ConstraintParams>> constraintPredicates;
+
+
+    public AllocationConstraints(ConstraintMode indexShardsPerNodeConstraint) {
+        this.indexShardsPerNodeConstraint = indexShardsPerNodeConstraint;
+        this.constraintPredicates = new ArrayList<>(1);
+        this.constraintPredicates.add(isIndexShardsPerNodeBreached());
+    }
+
+    public ConstraintMode getIndexShardsPerNodeConstraint() {
+        return indexShardsPerNodeConstraint;
+    }
+
+    class ConstraintParams {
+        private BalancedShardsAllocator.Balancer balancer;
+        private BalancedShardsAllocator.ModelNode node;
+        private String index;
+        private ConstraintMode invocationMode;
+
+        ConstraintParams(BalancedShardsAllocator.Balancer balancer, BalancedShardsAllocator.ModelNode node,
+                        String index, ConstraintMode invocationMode) {
+            this.balancer = balancer;
+            this.node = node;
+            this.index = index;
+            this.invocationMode = invocationMode;
+        }
+    }
+
+    /**
+     * Evaluates configured allocation constraint predicates for given node - index combination; and returns
+     * a weight value based on the number of breached constraints.
+     *
+     * Constraint weight should be added to the weight calculated via weight function, to reduce priority
+     * of allocating on nodes with breached constraints.
+     */
+    public long weight(BalancedShardsAllocator.Balancer balancer, BalancedShardsAllocator.ModelNode node,
+                       String index, ConstraintMode invocationMode) {
+        int constraintsBreached = 0;
+        ConstraintParams params = new ConstraintParams(balancer, node, index, invocationMode);
+        for (Predicate<ConstraintParams> predicate: constraintPredicates) {
+            if (predicate.test(params)) {
+                constraintsBreached++;
+            }
+        }
+        return constraintsBreached * CONSTRAINT_WEIGHT;
+    }
+
+    /**
+     * Constraint to control number of shards of an index allocated on a single node.
+     *
+     * In current weight function implementation, when a node has significantly fewer shards than other nodes (e.g. during
```

Comment:
I can see that this will break for use-cases where there are multiple indices in the cluster and all shards across all indices are equivalent in terms of resource usage. In that case, the existing weight function will turn out to be better. Also, the weight function becomes non deterministic now as this applies only when there are unassigned shards.

Reply:
Yes, let us add documentation for this as well.